File: solution/array/042-trapping-rain-water/solution.go
package main

func trap(height []int) int {
	n := len(height)
	if n < 3 {
		return 0
	}

	left, right := 0, n-1
	leftMax, rightMax := 0, 0
	trappedWater := 0

	for left < right {
		if height[left] > leftMax {
			leftMax = height[left]
		}

		if height[right] > rightMax {
			rightMax = height[right]
		}

		if leftMax < rightMax {
			trappedWater += leftMax - height[left]
			left++
		} else {
			trappedWater += rightMax - height[right]
			right--
		}
	}

	return trappedWater
}

File: solution/array/042-trapping-rain-water/solution.js
/**
 * @param {number[]} height
 * @return {number}
 */
function trap(height) {
    if (height.length < 3) {
        return 0;
    }

    let left = 0;
    let right = height.length - 1;

    let leftMax = 0;
    let rightMax = 0;

    let trappedWater = 0;

    while (left < right) {
        leftMax = Math.max(leftMax, height[left]);
        rightMax = Math.max(rightMax, height[right]);

        if (leftMax < rightMax) {
            trappedWater += leftMax - height[left];
            left++;
        } else {
            trappedWater += rightMax - height[right];
            right--;
        }
    }

    return trappedWater;
}